core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -0,0 +1,173 @@
+}
+
+
+class TransactionDataAndCallbacks(val transactionData: AddPartitionsToTxnTransactionCollection,
+                                  val callbacks: mutable.Map[String, AddPartitionsToTxnManager.AppendCallback])
+
+
+class AddPartitionsToTxnManager(config: KafkaConfig, client: NetworkClient, time: Time) 
+  extends InterBrokerSendThread("AddPartitionsToTxnSenderThread-" + config.brokerId, client, config.requestTimeoutMs, time) {
+  
+  private val inflightNodes = mutable.HashSet[Node]()
+  private val nodesToTransactions = mutable.Map[Node, TransactionDataAndCallbacks]()
+  
+  def addTxnData(node: Node, transactionData: AddPartitionsToTxnTransaction, callback: AddPartitionsToTxnManager.AppendCallback): Unit = {
+    // Check if we have already (either node or individual transaction). 
+    val currentNodeAndTransactionDataOpt = nodesToTransactions.get(node)
+    currentNodeAndTransactionDataOpt match {
+      case None =>
+        nodesToTransactions.put(node,
+          new TransactionDataAndCallbacks(new AddPartitionsToTxnTransactionCollection(Collections.singletonList(transactionData).iterator()),
+            mutable.Map(transactionData.transactionalId() -> callback)))
+      case Some(currentNodeAndTransactionData) =>
+        // Check if we already have txn ID -- this should only happen in epoch bump case. If so, we should return error for old entry and remove from queue.
+        val currentTransactionData = currentNodeAndTransactionData.transactionData.find(transactionData.transactionalId)
+        if (currentTransactionData != null) {
+          if (currentTransactionData.producerEpoch() < transactionData.producerEpoch()) {
+            val topicPartitionsToError = mutable.Map[TopicPartition, Errors]()
+            currentTransactionData.topics().forEach { topic => 
+              topic.partitions().forEach { partition =>
+                topicPartitionsToError.put(new TopicPartition(topic.name(), partition), Errors.INVALID_PRODUCER_EPOCH)
+              }
+            }
+            val oldCallback = currentNodeAndTransactionData.callbacks(transactionData.transactionalId())
+            currentNodeAndTransactionData.transactionData.remove(transactionData)
+            oldCallback(topicPartitionsToError.toMap)
+          } else {
+            // We should never see a request on the same epoch since we haven't finished handling the one in queue
+            throw new InvalidRecordException("Received a second request from the same connection without finishing the first.")
+          }
+        }
+        currentNodeAndTransactionData.transactionData.add(transactionData)
+        currentNodeAndTransactionData.callbacks.put(transactionData.transactionalId(), callback)
+    }
+    wakeup()
+  }
+
+  private class AddPartitionsToTxnHandler(node: Node, transactionDataAndCallbacks: TransactionDataAndCallbacks) extends RequestCompletionHandler {
+    override def onComplete(response: ClientResponse): Unit = {
```

Comment:
We should measure the time it took to process the request to TC and report it (either as "remote time" that we already have or as a new metric), otherwise it'll just roll into "local time" (time spent by this broker to handle the request).

Reply:
Yeah. I planned to look at the request timing stuff next. Thanks for the reminder :)

Reply:
Actually -- I misread the original comment -- I thought this was about callback time.
I'm wondering if we want to report this in a different way after all. It would be a bit confusing to report this as remote time since this is technically still on the broker.

I'm also not sure how we would parse the separate metric.

Is the idea that we would somehow get the time to send and receive the response and subtract it from the local time? 

I don't even see the local time metric getting incremented at this level. One idea is to take the request time and subtract the time to do the callback. I'd have to see if this is even possible though.

Reply:
Actually, this is a higher level discussion that we should probably have on the KIP thread, as it affects public metric semantics.  I think it would be useful to have a separate metric, because it'll help with diagnosing extra latency in transactions that we're introducing with this RPC call.  Especially if we provide a feature flag to turn off this check for perf reasons -- we'd need to have some data to help the admin make a decision.